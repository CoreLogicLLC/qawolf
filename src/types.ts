export type Action =
  | 'click'
  | 'fill'
  | 'goBack'
  | 'goForward'
  | 'goto'
  | 'press'
  | 'reload'
  | 'scroll'
  | 'selectOption'
  | 'type';

export type BrowserName = 'chromium' | 'firefox' | 'webkit';

export type Callback<S = void, T = void> = (data?: S) => T;

export interface Doc {
  attrs: Record<string, string>;
  name: string;
}

export interface PageEvent {
  frameIndex?: number;
  frameSelector?: string;
  name: string;
  page: number;
  time: number;
  value?: string | ScrollValue | null;
}

export interface WindowEvent extends PageEvent {
  name: WindowEventName;
}

export type WindowEventName =
  | 'goBack'
  | 'goForward'
  | 'goto'
  | 'reload';

export interface ElementEvent extends PageEvent {
  isTrusted: boolean;
  name: ElementEventName;
  selector: string;
  target: Doc;
}

export type ElementEventName =
  | 'change'
  | 'click'
  | 'input'
  | 'keydown'
  | 'keyup'
  | 'mousedown'
  | 'paste'
  | 'scroll';

export interface InputEvent extends ElementEvent {
  name: 'input';
  value: string | null;
}

export interface KeyEvent extends ElementEvent {
  name: 'keydown' | 'keyup';
  // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code
  value: string;
}

export interface LogEvent {
  level: string;
  message: string;
}

export interface PasteEvent extends ElementEvent {
  name: 'paste';
  value: string;
}

export interface ScrollEvent extends ElementEvent {
  name: 'scroll';
  value: ScrollValue;
}

export type ScrollValue = {
  x: number;
  y: number;
};

export interface Step {
  action: Action;
  event: PageEvent;
  // needed to build selector key
  index: number;
  value?: StepValue;
}

export type StepValue = string | ScrollValue | null | undefined;

export interface Workflow {
  device?: string;
  name: string;
  steps: Step[];
  url: string;
}
