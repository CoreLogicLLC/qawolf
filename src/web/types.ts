export type SelectorPart = {
  name: 'css' | 'text';
  body: string;
};

export type ParsedSelector = {
  capture?: boolean;
  parts: SelectorPart[];
};

export type Evaluator = {
  createTextSelector(element: Element): string | undefined;

  isVisible(element: Element): boolean;

  querySelector(selector: ParsedSelector, root: Node): HTMLElement;
};
