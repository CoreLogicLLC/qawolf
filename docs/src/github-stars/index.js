module.exports = function(context) {
  const { siteConfig } = context;
  const { themeConfig } = siteConfig;
  const { github } = themeConfig || {};

  if (!github || !github.repo) {
    throw new Error(`You need to specify 'github.repo' in 'themeConfig'`);
  }

  return {
    name: "github-stars",

    injectHtmlTags() {
      return {
        headTags: [
          {
            tagName: "script",
            innerHTML: `
            function htmlToElement(html) {
              var template = document.createElement('template');
              html = html.trim(); // Never return a text node of whitespace as the result
              template.innerHTML = html;
              return template.content.firstChild;
            }

            var starsIntervalId = setInterval(() => {
              const stars = document.querySelector(".github-wrapper");
              if (stars) return;

              const navBar = document.querySelector(".navbar__items.navbar__items--right");
              if (!navBar) return;

              var starsLink = htmlToElement('<div class="github-wrapper"><a class="github-button" href="https://github.com/${github.repo}" data-icon="octicon-star" data-size="large" data-show-count="true" aria-label="Star ${github.repo} on GitHub">Star</a></div>');
              navBar.append(starsLink);
              
              var starsScript = document.createElement('script');
              starsScript.src = "https://buttons.github.io/buttons.js";
              document.head.append(starsScript);
            }, 100);`
          }
        ]
      };
    }
  };
};
