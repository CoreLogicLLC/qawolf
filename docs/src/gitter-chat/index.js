module.exports = function(context) {
  const { siteConfig } = context;
  const { themeConfig } = siteConfig;
  const { gitter } = themeConfig || {};

  if (!gitter || !gitter.room) {
    throw new Error(`You need to specify 'gitter.room' in 'themeConfig'`);
  }

  return {
    name: "gitter-chat",

    injectHtmlTags() {
      return {
        headTags: [
          {
            tagName: "script",
            innerHTML: `
            ((window.gitter = {}).chat = {}).options = {
              room: '${gitter.room}'
            };
            `
          },
          {
            tagName: "script",
            attributes: {
              async: true,
              defer: true,
              src: "https://sidecar.gitter.im/dist/sidecar.v1.js"
            }
          }
        ]
      };
    }
  };
};
