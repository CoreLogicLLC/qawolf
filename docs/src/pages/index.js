function Home() {
  if (typeof window !== "undefined") {
    // redirect to Home page
    window.location.href = "https://www.qawolf.com";
  }

  return null;
}

export default Home;
