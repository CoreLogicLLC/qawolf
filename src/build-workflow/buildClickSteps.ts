import { isInputTarget } from './target';
import { ElementEvent, Step } from '../types';

const CLICK_EVENTS = ['click', 'mousedown'];

const shouldIncludeClickEvent = (
  event: ElementEvent,
  previousEvent?: ElementEvent,
): boolean => {
  // ignore system initiated clicks
  if (!event.isTrusted) return false;

  // ignore non-click events
  if (!CLICK_EVENTS.includes(event.name)) return false;

  // ignore clicks on selects
  if (event.target.name === 'select') return false;

  // skip system-initiated clicks triggered by a key press
  // ex. "Enter" triggers a click on a submit input
  if (
    previousEvent &&
    ['change', 'keydown', 'keyup'].includes(previousEvent.name) &&
    event.time - previousEvent.time < 50
  ) {
    return false;
  }

  return true;
};

const groupClickEvents = (events: ElementEvent[]): ElementEvent[][] => {
  const clickGroups = [];

  let group: ElementEvent[] = [];

  events.forEach((event, i) => {
    const previousEvent = events[i - 1];

    if (!shouldIncludeClickEvent(event, previousEvent)) return;

    // always group clicks with the previous mousedown/click events
    // a click will follow a mousedown if the mouse was released within the same element
    // a click will follow a click if it propagated to a higher element (a click on a label propagates to the input)
    if (!group.length || event.name === 'click') {
      group.push(event);
      return;
    }

    // start a new group
    clickGroups.push(group);
    group = [event];
  });

  if (group.length) {
    // append last group
    clickGroups.push(group);
  }

  return clickGroups;
};

export const buildClickSteps = (events: ElementEvent[]): Step[] => {
  const groupedClickEvents = groupClickEvents(events);
  const steps: Step[] = [];

  groupedClickEvents.forEach((events) => {
    let event = events[0] as ElementEvent;

    const inputEvent = events.find((event) => isInputTarget(event.target));
    if (inputEvent) {
      // if an event in the group is on an input, assume the click propagated
      // to an element like a checkbox or radio, which is most accurate target
      event = inputEvent;
    }

    steps.push({
      action: 'click',
      event,
      index: steps.length,
    });
  });

  return steps;
};
