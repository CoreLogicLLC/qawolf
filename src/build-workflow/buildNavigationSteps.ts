import { Action, WindowEvent, Step } from '../types';

export const buildNavigationSteps = (events: WindowEvent[]): Step[] => {
  return events.map((event, index) => ({
    action: (event.name as Action),
    event,
    index,
    value: event.value,
  }));
};
