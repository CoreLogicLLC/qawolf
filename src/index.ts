#!/usr/bin/env node

import * as qawolf from './qawolf';
import { Registry } from './utils';

// support: const qawolf = require("qawolf");
// support: import { ... } from "qawolf";
export * from './qawolf';

// support: import qawolf from "qawolf"
export default qawolf;

// must do this here to prevent circular dependency
Registry.instance().setQawolf(qawolf);

if (!module.parent) {
  // run the cli when this is the root module
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { runCli } = require('./cli/cli');
  runCli();
}
