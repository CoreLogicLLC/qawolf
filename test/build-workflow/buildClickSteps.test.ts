import { loadFixtures } from '../loadFixtures';
import { buildClickSteps } from '../../src/build-workflow';
import { ElementEvent } from '../../src/types';

describe('buildClickSteps', () => {
  const events: Record<string, ElementEvent[]> = {};

  beforeAll(async () => {
    await Promise.all(
      ['login', 'radioInputs', 'searchGoogle', 'selectNative'].map(
        async (name) => {
          events[name] = (await loadFixtures(name)).events;
        },
      ),
    );
  });

  it('builds one click per group of mousedown/click events', () => {
    const steps = buildClickSteps(events.login);
    expect(steps.map((step) => (step.event as ElementEvent).selector)).toMatchSnapshot();
  });

  it('groups [mousedown, change, click] sequence together', () => {
    const steps = buildClickSteps(events.searchGoogle);
    expect(steps.map((step) => (step.event as ElementEvent).selector)).toMatchSnapshot();
  });

  it('prefers clicks on inputs', () => {
    const steps = buildClickSteps(events.radioInputs);
    expect(steps.map((step) => (step.event as ElementEvent).selector)).toMatchSnapshot();
  });

  it('skips click triggered by Enter', () => {
    const steps = buildClickSteps(events.login);

    // click on login link
    expect((steps[0].event as ElementEvent).target.attrs.href).toEqual('/login');

    // click on initial input
    expect((steps[1].event as ElementEvent).selector).toEqual('#username');

    // click logout
    expect((steps[2].event as ElementEvent).selector).toEqual('text=Log out');
  });

  it('skips click on select', () => {
    const steps = buildClickSteps(events.selectNative);
    expect(steps.length).toBe(0);
  });
});
