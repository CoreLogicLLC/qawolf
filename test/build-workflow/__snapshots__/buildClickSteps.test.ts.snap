// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildClickSteps builds one click per group of mousedown/click events 1`] = `
Array [
  "text=Log in",
  "#username",
  "text=Log out",
]
`;

exports[`buildClickSteps groups [mousedown, change, click] sequence together 1`] = `
Array [
  "[aria-label=\\"Search\\"]",
  "[aria-label=\\"Google Search\\"]",
]
`;

exports[`buildClickSteps prefers clicks on inputs 1`] = `
Array [
  "[data-qa='html-radio']",
  "[data-qa='html-radio-group'] [value='dog']",
  "[data-qa='material-radio'] [value='yes']",
  "[data-qa='material-radio-group'] [value='green']",
]
`;
